core/src/main/java/org/ehcache/config/builders/ConfigurationBuilder.java

```diff
@@ -0,0 +1,208 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.unmodifiableCollection;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * The {@code ConfigurationBuilder} enables building {@link Configuration}s using a fluent style.
+ *
+ * @author Alex Snaps
+ */
+public class ConfigurationBuilder implements FluentConfigurationBuilder<ConfigurationBuilder> {
+
+  private final Map<String, CacheConfiguration<?, ?>> caches;
+  private final Collection<ServiceCreationConfiguration<?, ?>> serviceConfigurations;
+  private final ClassLoader classLoader;
+
+  public static ConfigurationBuilder newConfigurationBuilder() {
+    return new ConfigurationBuilder();
+  }
+
+  public static ConfigurationBuilder newConfigurationBuilder(Configuration seed) {
+    return new ConfigurationBuilder(new ConfigurationBuilder(new ConfigurationBuilder(new ConfigurationBuilder(),
+      seed.getCacheConfigurations()), seed.getServiceCreationConfigurations()), seed.getClassLoader());
+  }
+
+  private ConfigurationBuilder() {
+    this.caches = emptyMap();
+    this.serviceConfigurations = emptyList();
+    this.classLoader = null;
+  }
+
+  private ConfigurationBuilder(ConfigurationBuilder builder, Map<String, CacheConfiguration<?, ?>> caches) {
+    this.caches = unmodifiableMap(caches);
+    this.serviceConfigurations = builder.serviceConfigurations;
+    this.classLoader = builder.classLoader;
+  }
+
+  private ConfigurationBuilder(ConfigurationBuilder builder, Collection<ServiceCreationConfiguration<?, ?>> serviceConfigurations) {
+    this.caches = builder.caches;
+    this.serviceConfigurations = unmodifiableCollection(serviceConfigurations);
+    this.classLoader = builder.classLoader;
+  }
+
+  private ConfigurationBuilder(ConfigurationBuilder builder, ClassLoader classLoader) {
+    this.caches = builder.caches;
+    this.serviceConfigurations = builder.serviceConfigurations;
+    this.classLoader = classLoader;
+  }
+
+  @Override
+  public Configuration build() {
+    return new DefaultConfiguration(caches, classLoader, serviceConfigurations.toArray(new ServiceCreationConfiguration<?, ?>[serviceConfigurations.size()]));
+  }
+
+  @Deprecated
+  public ConfigurationBuilder addCache(String alias, CacheConfiguration<?, ?> config) {
+    CacheConfiguration<?, ?> existing = getCache(alias);
+    if (existing == null) {
+      return withCache(alias, config);
+    } else {
+      throw new IllegalArgumentException("Cache '" + alias + "' already exists: " + existing);
+    }
+  }
+
+  @Deprecated
+  public ConfigurationBuilder removeCache(String alias) {
```

Comment:
I presume that, even though this is a "new" class, it existed elsewhere before.  If there is existing non-internal code that uses this, it would benefit from having the `@deprecated` javadoc tag that indicates this method's replacement.

Reply:
Added javadoc.